### PR TITLE
inference: fix missing `LimitedAccuracy` handling in `conditional_changes`

### PR DESCRIPTION
**Pull request by aviatesk**
The following JET analysis told me that we missed to handle
`LimitedAccuracy` in `conditional_changes` and limited conditional
constraints weren't propagated:
```julia
julia> using JET, Test

julia> report_call() do
           @testset "foo" begin; true; end
       end
┌ @ /Users/aviatesk/julia/julia/usr/share/julia/stdlib/v1.7/Test/src/Test.jl:1262 Test.finish(ts)
│┌ @ /Users/aviatesk/julia/julia/usr/share/julia/stdlib/v1.7/Test/src/Test.jl:1007 Test.print_test_results(ts)
││┌ @ /Users/aviatesk/julia/julia/usr/share/julia/stdlib/v1.7/Test/src/Test.jl:941 #self#(ts, 0)
│││┌ @ /Users/aviatesk/julia/julia/usr/share/julia/stdlib/v1.7/Test/src/Test.jl:961 Test.get_alignment(ts, 0)
││││┌ @ /Users/aviatesk/julia/julia/usr/share/julia/stdlib/v1.7/Test/src/Test.jl:1034 Test.map(#36, Base.getproperty(ts, :results))
│││││┌ @ abstractarray.jl:2362 Base.collect_similar(A, Base.Generator(f, A))
││││││┌ @ array.jl:638 Base._collect(cont, itr, Base.IteratorEltype(itr), Base.IteratorSize(itr))
│││││││┌ @ array.jl:723 Base.iterate(itr)
││││││││┌ @ generator.jl:47 Base.getproperty(g, :f)(Base.getindex(y, 1))
│││││││││┌ @ /Users/aviatesk/julia/julia/usr/share/julia/stdlib/v1.7/Test/src/Test.jl:1034 Test.get_alignment(t, Test.+(Core.getfield(#self#, :depth), 1))
│││││││││││││┌ @ array.jl:728 Base.collect_to_with_first!(Base._similar_for(c, Base.typeof(v1), itr, isz), v1, itr, st)
││││││││││││││┌ @ array.jl:734 Base.collect_to!(dest, itr, Base.+(i1, 1), st)
│││││││││││││││┌ @ /Users/aviatesk/julia/julia/base/array.jl:764 Base.collect_to!(new, itr, Base.+(i, 1), st)
││││││││││││││││┌ @ /Users/aviatesk/julia/julia/base/array.jl:758 Base.indexed_iterate(y, 2, _8)
│││││││││││││││││┌ @ tuple.jl:94 Base.iterate(I, state)
││││││││││││││││││ no matching method found for call signature (Tuple{typeof(iterate), Nothing, Int64}): Base.iterate(I::Nothing, state::Int64)
│││││││││││││││││└───────────────
││││││││││││││││┌ @ /Users/aviatesk/julia/julia/base/array.jl:758 Base.indexed_iterate(y, 1)
│││││││││││││││││┌ @ tuple.jl:89 Base.iterate(I)
││││││││││││││││││ no matching method found for call signature (Tuple{typeof(iterate), Nothing}): Base.iterate(I::Nothing)
│││││││││││││││││└───────────────
Union{Test.FallbackTestSet, Test.DefaultTestSet}
```

After this PR:
```julia
julia> report_call() do
           @testset "foo" begin; true; end
       end
No errors !
Union{Test.FallbackTestSet, Test.DefaultTestSet}
```

Well, it's hard to test this kind of intermediate state of inference, so
I didn't include test cases this PR fixes.
(maybe can we use JET as an inference test ?)

### CODE REVIEW

**Review by vtjnash — COMMENTED** [review]
Might want to leave behind a comment about why this is legal though? The `⊑` was answering this query correctly, since the value `newtyp` is indeed lower in the lattice than `typ`, but we're using `⊑` as an approximation to taking the intersection (`tmeet`).

File: base/compiler/abstractinterpretation.jl [hunk 1]
```diff
@@ -1846,7 +1846,9 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
 end
 
 function conditional_changes(changes::VarTable, @nospecialize(typ), var::Slot)
-    if typ ⊑ (changes[slot_id(var)]::VarState).typ
+    newtyp = (changes[slot_id(var)]::VarState).typ
+    newtyp = ignorelimited(newtyp)
+    if typ ⊑ newtyp
```
[the remark this answers is not in the document]
Comment (aviatesk) [reply base/compiler/abstractinterpretation.jl:1851]: Right,  we should take into account the original "causes" here, although  the "causes" of `typ` and `newtyp` would be almost always same.

Also, I just found `newtyp` is completely misreading and so will rename it to `oldtyp`.

[the remark this answers is not in the document]
Comment (vtjnash) [reply base/compiler/abstractinterpretation.jl]: If they were the same, then the old `⊑` would have already taken that into account

File: base/compiler/abstractinterpretation.jl [hunk 2]
```diff
@@ -1847,8 +1847,11 @@ end
 
 function conditional_changes(changes::VarTable, @nospecialize(typ), var::Slot)
     newtyp = (changes[slot_id(var)]::VarState).typ
-    newtyp = ignorelimited(newtyp)
-    if typ ⊑ newtyp
+    # approximate test for `typ ∩ newtyp` being better than `newtyp`
+    # since we probably formed these types with `typesubstract`, the comparison is likely simple
+    if ignorelimited(typ) ⊑ ignorelimited(newtyp)
+        # typ is better unlimited, but we may still need to compute the tmeet with the limit "causes" since we ignored those in the comparison
+        newtyp isa LimitedAccuracy && (typ = tmerge(typ, LimitedAccuracy(Bottom, newtype.causes)))
```
[the remark this answers is not in the document]
Comment (aviatesk) [reply base/compiler/abstractinterpretation.jl]: 220abb1

**Review by vtjnash — APPROVED** [review]

**Comment by aviatesk** [comment]
@vtjnash bump

**Comment by vtjnash** [comment]
oh, oops, I assumed you would be able to merge

**Comment by aviatesk** [comment]
hehe, I don't have the write access yet.
I'd appreciate if you merge #40745 as well.

**Pull request merged**